bundles/framework/featuredata/plugin/FeaturedataPlugin.js

```diff
@@ -0,0 +1,149 @@
+         */
+        refresh: function () {
+            if (!this.handler) {
+                this.handler = new FeatureDataPluginHandler(this, this.getMapModule(), this.getConfig());
+            }
+            this.renderButton();
+            this.handler.getController().updateStateAfterMapEvent();
+        },
+        showLoadingIndicator: function (blnLoad) {
+            this.renderButton(!!blnLoad);
+        },
+        renderButton: function (loading = false) {
+            const el = this.getElement();
+            if (!el) {
+                return;
+            }
+
+            ReactDOM.render(
+                <ThemeProvider value={this.getMapModule().getMapTheme()}>
+                    <FeatureDataButton
+                        visible={this._hasFeaturedataLayers()}
+                        icon={<Message messageKey='title' bundleKey='FeatureData'/>}
+                        onClick={() => this.handler.openFlyout()}
+                        active={this._flyoutOpen}
+                        loading={loading}
+                        position={this.getLocation()}
+                    />
+                </ThemeProvider>,
+                el[0]
+            );
+        },
+        resetUI: function () {
+            if (this.handler) {
+                this.handler.closeFlyout();
+            }
+        },
+        /**
+         * @method _hasFeaturedataLayers
+         * @private
+         * Check whether there are layers with featuredata present -> determine the control element's visibility
+         */
+        _hasFeaturedataLayers: function () {
+            // see if there's any wfs layers, show element if so
+            return this.getSandbox()
+                .findAllSelectedMapLayers()
+                .filter(layer => layer.isVisibleOnMap())
+                .some(layer => layer.hasFeatureData && layer.hasFeatureData());
+        },
+        _createEventHandlers: function () {
+            return {
+                /**
+                 * @method AfterMapMoveEvent
+                 * Shows map center coordinates after map move
+                 */
+                AfterMapMoveEvent: function () {
+                    this.refresh();
+                },
+                /**
+                 * @method AfterMapLayerAddEvent
+                 * @param {Oskari.mapframework.event.common.AfterMapLayerAddEvent} event
+                 *
+                 * Calls flyouts layerAdded() method
+                 */
+                AfterMapLayerAddEvent: function (event) {
+                    if (event.getMapLayer().hasFeatureData()) {
+                        this.refresh();
+                    }
+                },
+                AfterMapLayerRemoveEvent: function (event) {
+                    if (event.getMapLayer().hasFeatureData()) {
+                        this.refresh();
+                    }
+                },
+                MapLayerVisibilityChangedEvent: function () {
+                    this.refresh();
```

Comment:
I would just call `this.handler.updateStateAfterMapEvent();` directly here since that should be the only thing we need. If there's no longer layers with featuredata -> state will be updated -> state change should trigger renderButton().